### PR TITLE
Split history tree insert from history node append

### PR DESCRIPTION
**Pull request by norberthu**
**What changed?**

**Why?**

**How did you test it?**

**Potential risks**

**Is hotfix candidate?**

### CODE REVIEW

File: common/persistence/history_manager.go [hunk 1]
```diff
@@ -372,14 +372,14 @@ func ToHistoryTreeInfo(serializer serialization.Serializer, blob *commonpb.DataB
 func (m *executionManagerImpl) serializeAppendHistoryNodesRequest(
 	ctx context.Context,
 	request *AppendHistoryNodesRequest,
-) (*InternalAppendHistoryNodesRequest, error) {
+) (*InternalInsertHistoryTreeRequest, *InternalAppendHistoryNodesRequest, error) {
```
Comment (MichaelSnowden) [inline common/persistence/history_manager.go:375]: Could you extract this to a separate method? I don't think the 3 return values look good here

File: common/persistence/history_manager.go [hunk 2]
```diff
@@ -435,7 +434,13 @@ func (m *executionManagerImpl) serializeAppendHistoryNodesRequest(
 		ShardID: request.ShardID,
 	}
 
-	if req.IsNewBranch {
+	if nodeID < GetBeginNodeID(branch) {
```
Comment (MichaelSnowden) [inline common/persistence/history_manager.go:437]: Why is this moved up?

File: common/persistence/history_manager.go [hunk 3]
```diff
@@ -444,54 +449,51 @@ func (m *executionManagerImpl) serializeAppendHistoryNodesRequest(
...
 
-	return req, nil
+	return nil, req, nil
 }
 
 func (m *executionManagerImpl) serializeAppendRawHistoryNodesRequest(
 	ctx context.Context,
 	request *AppendRawHistoryNodesRequest,
-) (*InternalAppendHistoryNodesRequest, error) {
+) (*InternalInsertHistoryTreeRequest, *InternalAppendHistoryNodesRequest, error) {
```
Comment (MichaelSnowden) [inline common/persistence/history_manager.go:467]: Ditto about the 3 return values.

**Review by MichaelSnowden — COMMENTED** [review]
To confirm, this is just for refactoring to make things cleaner, and there are no intended behavioral changes?

**Review by norberthu — COMMENTED** [review]
Correct, the logical behavior should be identical. Also, I took your suggestion to break the serializeXXX helpers up and therefore the comment about moving the if-statement is moot.

File: common/persistence/history_manager.go [hunk 4]
```diff
@@ -532,16 +528,23 @@ func (m *executionManagerImpl) AppendHistoryNodes(
...
 
-	req, err := m.serializeAppendHistoryNodesRequest(ctx, request)
-
+	nodeReq, err := m.serializeAppendHistoryNodesRequest(request)
 	if err != nil {
 		return nil, err
 	}
 
-	err = m.persistence.AppendHistoryNodes(ctx, req)
+	err = m.persistence.AppendHistoryNodes(ctx, nodeReq)
```
Comment (MichaelSnowden) [inline common/persistence/history_manager.go:536]: Append and Insert used to be in the same batch. Are you sure this won't cause any problems with them being in separate batches? For example, what happens if we crash after 536 and before 542?

File: common/persistence/history_manager.go [hunk 5]
```diff
@@ -532,16 +528,23 @@ func (m *executionManagerImpl) AppendHistoryNodes(
...
+	if err == nil && request.IsNewBranch {
```
Comment (MichaelSnowden) [inline common/persistence/history_manager.go:537]: Why aren't we using the `if err != nil { return err }` pattern here?

Comment (MichaelSnowden) [reply common/persistence/history_manager.go:537]: And below

File: common/persistence/history_manager.go (hunk 4 above)
Comment (norberthu) [reply common/persistence/history_manager.go:536]: This failure condition (with nodes being written first with tree possibly not) is well known. The periodic scavenger task cleans these up.

File: common/persistence/history_manager.go (hunk 5 above)
Comment (norberthu) [reply common/persistence/history_manager.go:537]: It did strike me as a bit odd why the `size` is returned regardless of error or not. But I believe that's why the original code also had this slight anti-pattern for error handling. I tracked this entire callstack chain and as far as I can tell, the size is expected. Hence, I'm following suit.

Comment (MichaelSnowden) [reply common/persistence/history_manager.go:537]: I don't think we should repeat this pattern here because it sticks out like a sore thumb.

**Review by MichaelSnowden — APPROVED** [review]
Just a nit. LGTM otherwise
